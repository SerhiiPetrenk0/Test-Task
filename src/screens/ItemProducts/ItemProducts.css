.itemImg {
    height: 500px!important;
}