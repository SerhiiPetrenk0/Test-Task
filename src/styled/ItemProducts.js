import styled from 'styled-components';
import { Container, Card, Col } from 'react-bootstrap';

const StyledContainer = styled(Container)`
    margin-top: 50px
`;
const ColImg = styled(Col)`
    display: flex;
    justify-content: center
`;
const CImg = styled(Card)`
    display: block;
    justify-content: center;
    width: 30rem;
    height: 30rem;
`;
const CardImg = styled(Card.Img)`
    object-fit: contain;
    padding: 10px;
    max-width: 100%;
    max-height: 100%;
`;
const StyledCard = styled(Card)`
    width: 100%
`;


export { 
    StyledContainer as Container, 
    ColImg, 
    CImg, 
    CardImg, 
    StyledCard as Card
};
