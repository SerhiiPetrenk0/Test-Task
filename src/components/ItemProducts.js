import React, { useEffect, useState } from 'react';
import { Card, Row, Col } from 'react-bootstrap';
import { useParams } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import { PostForm, Comment } from './';

import { loadProduct, loadComment } from '../redux/ducks/productsDuck';
import * as Styled from '../styled/ItemProducts';
import { PageSpinner } from '../styled/globalStyled';

export const ItemProducts = props => {
    const [ itemProduct, setItemProduct ] = useState([]);
    const [ itemComment, setItemComment] = useState([]);
    const { products, comments } = useSelector(store => store.product);
    const status = useSelector(store => store.loader.status);
    const dispatch = useDispatch();
    const linkProduct = useParams();
    const checkStore = () => products.length === 0 && dispatch(loadProduct());

    const chooseProduct = () => {
        const chooseItem = products.find(item => item.asin === linkProduct.id);
        setItemProduct(chooseItem);
    };
    const checkComments = () => ( comments.length === 0 || comments.asin !== linkProduct.id) ? 
        dispatch(loadComment(linkProduct.id)) : 
        setItemComment(comments.body);
    
    const renderComment = itemComment.map((item, index) => <Comment key={index} ListComment={item} />);

    const deeps = [products, linkProduct.id, comments.body];
    useEffect(() => {
        checkComments();
        checkStore();
        chooseProduct();
    },[...deeps]);

    return (
        <>
            {!status ? ( 
                <PageSpinner animation="border" /> 
            ) : (
              <Styled.Container>
                  {itemProduct && 
                    <Row>
                        <Styled.ColImg lg="6">
                            <Styled.CImg>
                                <Styled.CardImg src={itemProduct.img} />
                            </Styled.CImg>
                        </Styled.ColImg>
                        <Col>
                            <Styled.Card>
                            <Card.Body>
                                <Card.Title>{itemProduct.brand}</Card.Title>
                                <Card.Text>
                                {itemProduct.name}
                                </Card.Text>
                                {renderComment} 
                            </Card.Body>
                            </Styled.Card>
                            <PostForm />
                        </Col>
                    </Row>
                  }
              </Styled.Container>
            )}
        </>
    );
};
