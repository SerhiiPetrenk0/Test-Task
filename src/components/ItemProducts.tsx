import React, { useEffect, useState } from 'react';
import { Card, Row, Col } from 'react-bootstrap';
import { useParams } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import { PostForm, Comment } from '.';

import { loadProduct, loadComment } from '../redux/ducks/productsDuck';
import * as Styled from '../styled/ItemProducts';
import { PageSpinner } from '../styled/globalStyled';
import { TypeProduct, TypeComment, TypePostForm } from '../interface';

export const ItemProducts: React.FC = () => {
    const [ itemProduct, setItemProduct ] = useState<TypeProduct | undefined>(undefined);
    const [ itemComment, setItemComment] = useState<TypePostForm[]>([]);
    type TProductsStore = {
        product: {
            products: TypeProduct[],
            comments: TypeComment
        }
    };
    type TStatusStore = {
        loader: {
            status: boolean
        }
    };
    const { products, comments } = useSelector((store: TProductsStore) => store.product);
    const status = useSelector((store: TStatusStore) => store.loader.status);
    const dispatch = useDispatch();
    const linkProduct = useParams();
    const checkStore = () => products.length === 0 && dispatch(loadProduct());
    const chooseProduct = () => {
        const chooseItem = products.find((item: { asin: string }) => item.asin === linkProduct.id);
        setItemProduct(chooseItem);
    };
    const checkComments = () => ( !comments.body || comments.asin !== linkProduct.id ) ? 
        dispatch(loadComment(linkProduct.id)) : 
        setItemComment(comments.body);
    
    const renderComment: JSX.Element[] = itemComment.map(
        (item: TypePostForm, index: number) => <Comment key={index} ListComment={item} />
    );

    useEffect(() => {
        checkComments();
        checkStore();
        chooseProduct();
    });

    return (
        <>
            {status ? ( 
                <PageSpinner animation="border" /> 
            ) : (
              <Styled.Container>
                  {itemProduct && 
                    <Row id='12dw3456789'>
                        <Styled.ColImg lg="6">
                            <Styled.CImg>
                                <Styled.CardImg src={itemProduct.img} />
                            </Styled.CImg>
                        </Styled.ColImg>
                        <Col>
                            <Styled.Card>
                            <Card.Body>
                                <Card.Title>{itemProduct.brand}</Card.Title>
                                <Card.Text>
                                {itemProduct.name}
                                </Card.Text>
                                {renderComment} 
                            </Card.Body>
                            </Styled.Card>
                            <PostForm />
                        </Col>
                    </Row>
                  }
              </Styled.Container>
            )}
        </>
    );
};
