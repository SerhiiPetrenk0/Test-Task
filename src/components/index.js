//Screens
import { ListProducts } from './ListProducts';
import { ItemProducts } from './ItemProducts';
//Components
import { Comment } from './components/Comment';
import { PostForm } from './components/PostForm';
import { StarBar } from './components/StarBar';

export { 
    ListProducts,
    ItemProducts, 
    Comment, 
    PostForm, 
    StarBar 
};
