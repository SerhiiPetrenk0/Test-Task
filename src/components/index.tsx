//Screens
import { ListProducts } from './ListProducts';
import { ItemProducts } from './ItemProducts';
//Components
import { Comment } from './other/Comment';
import { PostForm } from './other/PostForm';
import { StarBar } from './other/StarBar';

export { 
    ListProducts,
    ItemProducts, 
    Comment, 
    PostForm, 
    StarBar 
};
