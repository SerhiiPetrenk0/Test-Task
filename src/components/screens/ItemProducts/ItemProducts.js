import React, { useEffect, useState } from 'react';
import { Card, Row, Col } from 'react-bootstrap';
import { Comment } from '../components/Comment';
import { PostForm } from '../components/PostForm';
import { useParams } from 'react-router-dom';
import { Product } from '../../redux/Redux-store'; // Тестовий варіант >>> данні які будуть завантажуватись з REDUX
import { 
    StyledContainer,
    StyledColImg,
    StyledCImg,
    StyledCardImg,
    StyledCard,
 } from '../../../styled/screens/ItemProducts/ItemProducts';
 
const ListComment = [
    {
        id: 1,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 2,
    },
    {
        id: 2,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 5,
    },
    {
        id: 3,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 4,
    },
]

export const ItemProducts = (props) => {
    const [ itemProduct, setItemProduct ] = useState([])
    const linkProduct = useParams()

    const chooseProduct = () => {
        const chooseItem = Product.products.find(item => item.asin === linkProduct.id)
        setItemProduct(chooseItem)
    }

    const renderComment = ListComment.map(item => <Comment key={item.id} ListComment={item} />)

    useEffect(() => {
        chooseProduct();
    },[]);
    
    return(
        <StyledContainer>
            {itemProduct && 
                <Row /*className="flex-column flex-lg-row justify-content-between"*/>
                    <StyledColImg lg="6">
                        <StyledCImg>
                            <StyledCardImg src={itemProduct.img} />
                        </StyledCImg>
                    </StyledColImg>
                    <Col>
                        <StyledCard>
                        <Card.Body>
                            <Card.Title>{itemProduct.brand}</Card.Title>
                            <Card.Text>
                            {itemProduct.name}
                            </Card.Text>
                            {renderComment}
                        </Card.Body>
                        </StyledCard>
                        <PostForm />
                    </Col>
                </Row>
            }
        </StyledContainer>
    )
}