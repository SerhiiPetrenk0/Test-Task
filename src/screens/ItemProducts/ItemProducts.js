import React from "react";
import { Image, Card, Container, Row, Col } from "react-bootstrap";
import { Comment } from "../components/Commetn";

const ListComment = [
    {
        id: 1,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 2,
    },
    {
        id: 2,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 5,
    },
    {
        id: 3,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 4,
    },
    {
        id: 4,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 1,
    },
    {
        id: 5,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 3,
    },
    {
        id: 6,
        discription: "Some quick example text to build on the card title and make up the bulk of the card's content.",
        stars: 2,
    },
]

export const ItemProducts = (props) => {
    const { item } = props

    const renderComment = ListComment.map(item => <Comment ListComment={item} />)
    return(
        <Container className="mt-5">
            <Row>
                <Col>
                    <Image src={item.img} className="rounded-3" alt="***" />
                </Col>
                <Col>
                    <Card style={{ width: '100%' }}>
                    <Card.Body>
                        <Card.Title>{item.brand}</Card.Title>
                        <Card.Text>
                        {item.name}
                        </Card.Text>
                        {renderComment}
                    </Card.Body>
                    </Card>
                </Col>
            </Row>
        </Container>
    )
}