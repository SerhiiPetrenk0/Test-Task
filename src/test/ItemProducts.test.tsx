import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import * as reactRedux from 'react-redux'
import 'jsdom-global/register';

import { ItemProducts } from '../components/ItemProducts';

import { MockProductItem } from './MockData';

configure({ adapter: new Adapter() });

describe('ItemProducts component', () => {
  beforeEach(() => {
    const dummyDispatch = jest.fn();
    const myInitialState = MockProductItem;
    React.useState = jest.fn().mockReturnValue([myInitialState, {}]);
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    useSelectorMock.mockReturnValue(MockProductItem);
    useDispatchMock.mockReturnValue(dummyDispatch);
  });

  it('should match a snapshot (ItemProducts.test.tsx.snap)', () => {
    const component = shallow(<ItemProducts />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
