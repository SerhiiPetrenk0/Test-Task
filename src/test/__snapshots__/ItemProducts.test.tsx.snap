// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemProducts component should match a snapshot (ItemProducts.test.tsx.snap) 1`] = `
<Fragment>
  <Styled(Spinner)
    animation="border"
  />
</Fragment>
`;
